### ISSUE TITLE
Extracting the PLS-DA emmedding function 

### ISSUE THREAD

**Issue by sbrn3** [mixOmicsTeam/mixOmics#50]
Hi, I am currently trying to use the PLS-DA function for dimensionality reduction (not classification) and am therefore just interested in extracting the components that it generates. I was wondering if there is an easy way to project new test data into this reduced dimension embedding (the embedding is based only on the training data).

**Comment by mixOmicsTeam** [comment]
Hi Sam
With plotIndiv(style = graphics’) you can plot the samples on the training set, then overlay additional predicted coordinates using points().
To obtain the predicted coordinates, use the predict.plsda function (output called: variates).

Alternatively you can also do everything by hand and extract the components (variates) from a splsda object ($variates)

X <- breast.tumors$gene.exp
> # Y will be transformed as a factor in the function,
> # but we set it as a factor to set up the colors.
> Y <- as.factor(breast.tumors$sample$treatment)

> res <- splsda(X, Y, ncomp = 2, keepX = c(25, 25))
> res$variates$X



Regards
Kim-Anh
--
Dr. Kim-Anh Lê Cao
Snr Lecturer, NHMRC Career Development Fellow
School of Mathematics and Statistics, Melbourne Integrative Genomics Bld 184 | The University of Melbourne
T: +61 (0)3834 43971 | http://mixomics.org |@mixOmics_team | http://lecao-lab.science.unimelb.edu.au/














On 31 Oct 2019, at 3:10 pm, Sam <notifications@github.com<mailto:notifications@github.com>> wrote:

Hi, I am currently trying to use the PLS-DA function for dimensionality reduction (not classification) and am therefore just interested in extracting the components that it generates. I was wondering if there is an easy way to project new test data into this reduced dimension embedding (the embedding is based only on the training data).

**Comment by sbrn3** [comment]
Thankyou!

**Comment by sbrn3** [comment]
Hi, 

Really sorry to bother you again but I have a follow up question. So I am using the predict.plsda function to find the variates of a test set. It works fine when I do this across a whole tissue but when I subsample the data to only do it for one cell type, the test set returns all NaNs. 

1. I run PLS-DA for a training set of all the cells in a tissue
2. I use predict.plsda() and the output of the previous plsda to find the **variates** of my test set (cells from the same tissue type)
This works and I can get the variates of all the test cells. 

I found that the data was bit noisy so I decided to restrict this to one cell type. 
1. I run PLS-DA for a training set of one cell type. 
2. I use predict.plsda() and the output of the previous plsda to find the variates of my test set (subsample of test set which is the same cell type as the training set).
This gives me only NaNs for the variates. Although I am not interested in this information it might be relevant that all of the Majority vote values are empty strings (""). 

I'm not too sure what is going wrong here. Even when I subset to just one cell type there are more than 10 000 cells.

**Comment by frohart** [comment]
Hi Sam,

What is your outcome when you're running the PLS-DA? The cell type?
My guess would be that after reducing your data, you have features that are constant and that would mess up the prediction. 
Could you try to run the plsda model with a "near.zero.var=TRUE"?

**Comment by sbrn3** [comment]
Thanks! This works. 

On another note if you were interested there is a typo in the description of nearZeroVar: 
> zero variance predictors) or predictors that _are_ have both of the following characteristics
